// guardz-generator/dist/application/services/file-system/file-system.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSystemService = void 0;
class FileSystemService {
    fileSystem;
    constructor(fileSystem) {
        this.fileSystem = fileSystem;
    }
    async writeFilesToSameDirectory(files) {
        // Convert GeneratedFile[] to the format expected by IFileSystem
        const fileArray = files.map(file => ({
            fileName: file.fileName,
            content: file.content,
        }));
        await this.fileSystem.writeFilesToSameDirectory(fileArray);
    }
    async writeFilesToDirectory(files, baseOutputDir) {
        // Convert GeneratedFile[] to the format expected by IFileSystem
        const fileArray = files.map(file => ({
            fileName: file.fileName,
            content: file.content,
        }));
        await this.fileSystem.writeFilesToDirectory(fileArray, baseOutputDir);
    }
    async fileExists(filePath) {
        return this.fileSystem.fileExists(filePath);
    }
}
exports.FileSystemService = FileSystemService;
