// guardz-generator/dist/infrastructure/adapters/file-system/file-generation.adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileGenerationAdapter = void 0;
const file_1 = require("../../../core/generators/file");
const interface_1 = require("../../../core/generators/type-guard/interface");
const type_alias_1 = require("../../../core/generators/type-guard/type-alias");
class FileGenerationAdapter {
    generateInterfaceFile(interfaceDecl, guardName, outputDir, sourceFiles, allInterfaces, importMap) {
        return (0, file_1.generateInterfaceFile)(interfaceDecl, guardName, outputDir, sourceFiles, allInterfaces, importMap);
    }
    generateTypeAliasFile(typeAliasDecl, guardName, outputDir, sourceFiles, importMap) {
        return (0, file_1.generateTypeAliasFile)(typeAliasDecl, guardName, outputDir, sourceFiles, importMap);
    }
    generateCompleteFile(guardName, typeGuardCode, typeName, importTypes, fileName, outputDir, sourceFiles, hasGenerics, importMap, sourceFile) {
        return (0, file_1.generateCompleteFile)(guardName, typeGuardCode, typeName, importTypes, fileName, outputDir, sourceFiles, hasGenerics, importMap, sourceFile);
    }
    generateInterfaceTypeGuard(interfaceDecl, guardName, allInterfaces, sourceFiles, importMap) {
        return (0, interface_1.generateInterfaceTypeGuard)(interfaceDecl, guardName, allInterfaces, sourceFiles, importMap);
    }
    generateGenericInterfaceTypeGuard(interfaceDecl, guardName, allInterfaces, sourceFiles, importMap) {
        return (0, interface_1.generateGenericInterfaceTypeGuard)(interfaceDecl, guardName, allInterfaces, sourceFiles, importMap);
    }
    generateTypeAliasTypeGuard(typeAliasDecl, guardName, sourceFiles) {
        return (0, type_alias_1.generateTypeAliasTypeGuard)(typeAliasDecl, guardName, sourceFiles);
    }
}
exports.FileGenerationAdapter = FileGenerationAdapter;
