// guardz/dist/utils/validationUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validation = exports.reportValidationResults = exports.validateObject = exports.validateProperty = exports.createSimplifiedTree = exports.combineResults = exports.createTreeNode = exports.createValidationError = exports.createValidationResult = exports.getExpectedTypeName = void 0;
var getExpectedTypeName_1 = require("./getExpectedTypeName");
Object.defineProperty(exports, "getExpectedTypeName", { enumerable: true, get: function () { return getExpectedTypeName_1.getExpectedTypeName; } });
var createValidationResult_1 = require("./createValidationResult");
Object.defineProperty(exports, "createValidationResult", { enumerable: true, get: function () { return createValidationResult_1.createValidationResult; } });
var createValidationError_1 = require("./createValidationError");
Object.defineProperty(exports, "createValidationError", { enumerable: true, get: function () { return createValidationError_1.createValidationError; } });
var createTreeNode_1 = require("./createTreeNode");
Object.defineProperty(exports, "createTreeNode", { enumerable: true, get: function () { return createTreeNode_1.createTreeNode; } });
var combineResults_1 = require("./combineResults");
Object.defineProperty(exports, "combineResults", { enumerable: true, get: function () { return combineResults_1.combineResults; } });
var createSimplifiedTree_1 = require("./createSimplifiedTree");
Object.defineProperty(exports, "createSimplifiedTree", { enumerable: true, get: function () { return createSimplifiedTree_1.createSimplifiedTree; } });
var validateProperty_1 = require("./validateProperty");
Object.defineProperty(exports, "validateProperty", { enumerable: true, get: function () { return validateProperty_1.validateProperty; } });
var validateObject_1 = require("./validateObject");
Object.defineProperty(exports, "validateObject", { enumerable: true, get: function () { return validateObject_1.validateObject; } });
var reportValidationResults_1 = require("./reportValidationResults");
Object.defineProperty(exports, "reportValidationResults", { enumerable: true, get: function () { return reportValidationResults_1.reportValidationResults; } });
// Import all functions to create the Validation object
const createValidationResult_2 = require("./createValidationResult");
const combineResults_2 = require("./combineResults");
const createValidationError_2 = require("./createValidationError");
const createTreeNode_2 = require("./createTreeNode");
const validateProperty_2 = require("./validateProperty");
const validateObject_2 = require("./validateObject");
const reportValidationResults_2 = require("./reportValidationResults");
const createSimplifiedTree_2 = require("./createSimplifiedTree");
// Export the Validation object for backward compatibility
exports.Validation = {
    result: createValidationResult_2.createValidationResult,
    combine: combineResults_2.combineResults,
    error: createValidationError_2.createValidationError,
    treeNode: createTreeNode_2.createTreeNode,
    property: validateProperty_2.validateProperty,
    object: validateObject_2.validateObject,
    report: reportValidationResults_2.reportValidationResults,
    createSimplifiedTree: createSimplifiedTree_2.createSimplifiedTree
};
