// guardz-generator/dist/infrastructure/implementations/file-system/file-system.implementation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSystemImplementation = void 0;
const fs = __importStar(require("fs/promises"));
const path = __importStar(require("path"));
class FileSystemImplementation {
    async readFile(filePath) {
        return fs.readFile(filePath, 'utf-8');
    }
    async writeFile(filePath, content) {
        await fs.writeFile(filePath, content, 'utf-8');
    }
    async writeFilesToSameDirectory(files) {
        for (const file of files) {
            const filePath = path.join(file.fileName);
            await fs.writeFile(filePath, file.content, 'utf-8');
        }
    }
    async writeFilesToDirectory(files, baseOutputDir) {
        await fs.mkdir(baseOutputDir, { recursive: true });
        for (const file of files) {
            const filePath = path.join(baseOutputDir, file.fileName);
            await fs.writeFile(filePath, file.content, 'utf-8');
        }
    }
    async fileExists(filePath) {
        try {
            await fs.access(filePath);
            return true;
        }
        catch {
            return false;
        }
    }
    async mkdir(dirPath, options) {
        await fs.mkdir(dirPath, options);
    }
}
exports.FileSystemImplementation = FileSystemImplementation;
